File: 003/solution_003.ml
let init_divisor = 3

let rec is_divisible = function
    | n, divisor when n <= divisor -> false
    | n, divisor when (n mod divisor == 0) -> true
    | n, divisor -> is_divisible (n, (divisor + 2))

let is_prime = function
    | 0 -> false
    | 1 -> false
    | 2 -> true
    | n when (n mod 2 == 0) -> false
    | n when is_divisible (n, init_divisor) -> false
    | _ -> true

let rec largest_prime_factor = function
    | _, candidate when candidate <= 0 -> (false, 0)
    | n, candidate when (n mod candidate == 0) && (is_prime candidate) ->
      (true, candidate)
    | n, candidate -> largest_prime_factor (n, (candidate - 2))

let round n =
    int_of_float (floor (n +. 0.5))

let ensure_odd = function
    | n when n mod 2 == 0 -> (n + 1)
    | n -> n

let project_euler_003 n =
    let init_candidate = ensure_odd (round (sqrt (float_of_int n))) in
    match largest_prime_factor (n, init_candidate) with
    | true, solution -> print_newline (print_int solution)
    | false, _       -> print_newline (print_string "NO SOLUTION")

let () = project_euler_003 600851475143

File: 003/solution_003.js
#! /usr/bin/env node

"use strict"

var is_prime = function(n) {
    var is_prime = true;
    if (n != 2 && (n < 2 || n % 2 === 0)) {
        is_prime = false
    } else {
        var max_divisor = Math.round(Math.sqrt(n));
        var i = 3;
        for (i; i < max_divisor; i += 2) {
            if (n % i === 0) {
                is_prime = false;
                break
            }
        }
    }
    return is_prime
}

var project_euler_003 = function(n) {
    var largest_prime_factor = null;
    var candidate = Math.round(Math.sqrt(n / 2));
    candidate = (candidate % 2 === 0) ? + 1 : candidate;
    for (candidate; candidate > 1; candidate -= 2) {
        if (is_prime(candidate) && (n % candidate === 0)) {
            largest_prime_factor = candidate;
            break
        }
    }
    console.log("%d", largest_prime_factor)
}

project_euler_003(600851475143)
